src/DataLogger/DataLogger.cpp

```diff
@@ -0,0 +1,248 @@
+        }
+
+        // determine the number of columns
+        if (commaNum > 0) {
+            columnsInFile = commaNum + 1;
+        }
+        else if (charactersPresent) {
+            columnsInFile = 1;
+        }
+        else {
+            columnsInFile = 0;
+        }
+
+        DEBUG_SERIAL_LN("Counted " + (String)columnsInFile + " columns and reached end of row.");
+
+        // check if the number of columns in the file match the number of columns specified
+        if (columnsInFile == numColumns) {
+            // file is good
+            _numColumns = numColumns;
+            _curFilePtr = &tempFile;
+            return true;
+        }
+        // check if it is an empty file
+        else if (_numColumns == 0 && escaped == false)
+        {
+            // file is empty
+            _numColumns = numColumns;
+            _curFilePtr = &tempFile;
+            return true;
+        }
+        else 
+        {
+            // file is not empty, and does not have the correct number of columns
+            DEBUG_SERIAL_LN("Incorrect column number. Creating in file copyof_" + name);
+            tempFile.close();
+
+            // create new name for file
+
+            // find extension (example: myFile.csv)
+            int extensionIndex = name.lastIndexOf('.');
+            String base = name.substring(0, extensionIndex == -1 ? name.length() : extensionIndex);
+            String extension = extensionIndex == -1 ? "" : name.substring(extensionIndex + 1);
+            base.trim();
+
+            // find current file count (example: myFile (2).csv)
+            int fileCountStart = base.lastIndexOf('(');
+            int fileCountEnd = base.lastIndexOf(')');
+            int count = 0;
+            if (fileCountStart != -1 && fileCountEnd != -1 && fileCountEnd == base.length() - 1) {
+                String countString = base.substring(fileCountStart + 1, fileCountEnd);
+                bool numeric = true;
+                
+                // checks if all characters in countString are numeric
+                for (size_t i = 0; i < countString.length(); i++) {
+                    if (!isdigit(countString.charAt(i))) {
+                        numeric = false;
+                        break;
+                    }
+                }
+
+                if (numeric) {
+                    count = countString.toInt();
+                    // take count out of file name
+                    base = base.substring(0, fileCountStart);
+                    base.trim();
+                }
+            }
+            String newFileName = base + " (" + (count + 1) + ")" + (extension != "" ? "." + extension : "");
+            return open(newFileName, numColumns);
+        }
+    }
+    else
+    {
+        // file doesnt exist. create new file
+
+        DEBUG_SERIAL_LN(name + " doesn't exist. Creating new file");
+        return create(name, numColumns);
+    }
+}
+
+bool DataLogger::saveToDisk() {
+    if (_curFilePtr != nullptr)
```

Comment:
Considering the changes I've recommended above, if you want to check that there is a currently open valid file before you go to save, you can maintain a bool value (```DataLogger::_fileValid```) which indicates whether or not ```DataLogger::_curFile``` is valid

Reply:
Good idea, I have implemented this change

Reply:
Copying the same comment from above:

> Apparently the File class overloads the bool() operator so you can treat it as a nullable object:

```
if (_curFile) {
  // valid
} else {
  // invalid
}
```

> This is kind of strange, but but it means that you don't need to maintain that bool.